### PR TITLE
Arachne 1895/Multi-tenancy

### PR DESCRIPTION
**Pull request by pavgra**

### CODE REVIEW

File: src/main/java/com/odysseusinc/arachne/portal/config/tenancy/TenantContext.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
+package com.odysseusinc.arachne.portal.config.tenancy;
+
+import org.apache.commons.lang3.ObjectUtils;
+
+public class TenantContext {
+
+    private static ThreadLocal<Long> currentUserId = new ThreadLocal<>();
```
Comment (wivern) [inline src/main/java/com/odysseusinc/arachne/portal/config/tenancy/TenantContext.java]: unsync is possible here

File: src/main/resources/application-base.yml [hunk 2]
```diff
@@ -93,6 +93,8 @@ arachne:
   solrServerUrl: http://localhost:8983/solr
   mail:
     notifier: odysseusinc.notifier@gmail.com
+  tenancy:
+    currentTenantDbVariable: app.tenant_id
```
Comment (mrAlexandr) [inline src/main/resources/application-base.yml]: It is not needed becouse app.tenant_id is hardcoded in migration file

File: src/main/java/com/odysseusinc/arachne/portal/api/v1/controller/ExceptionHandlingController.java [hunk 3]
```diff
@@ -133,7 +133,6 @@
     @ExceptionHandler(AccessDeniedException.class)
     public ResponseEntity<JsonResult> exceptionHandler(AccessDeniedException ex) {
 
-        LOGGER.error(ex.getMessage(), ex);
```
Comment (mrAlexandr) [inline src/main/java/com/odysseusinc/arachne/portal/api/v1/controller/ExceptionHandlingController.java]: return back as LOGGER.error(ex.getMessage());

File: src/main/java/com/odysseusinc/arachne/portal/config/tenancy/HibernateConfig.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+@Configuration
+public class HibernateConfig {
+
+    @Autowired
+    private JpaProperties jpaProperties;
+
+    @Bean
+    public JpaVendorAdapter jpaVendorAdapter() {
+        return new HibernateJpaVendorAdapter();
```
Comment (mrAlexandr) [inline src/main/java/com/odysseusinc/arachne/portal/config/tenancy/HibernateConfig.java]: checkstyle (blank line should be below signature)

File: src/main/java/com/odysseusinc/arachne/portal/model/DataSource.java [hunk 5]
```diff
@@ -238,4 +249,14 @@ public String getOrganization() {
     public void setOrganization(String organization) {
         this.organization = organization;
```
Comment (mrAlexandr) [inline src/main/java/com/odysseusinc/arachne/portal/model/DataSource.java]: checkstyle (blank line should be below signature)

File: src/main/java/com/odysseusinc/arachne/portal/service/impl/BaseArachneSecureServiceImpl.java [hunk 6]
```diff
@@ -92,13 +96,24 @@ public BaseArachneSecureServiceImpl(UserStudyGroupedRepository userStudyGroupedR
...
     }
 
     @Override
     @Transactional(readOnly = true)
     public List<ParticipantRole> getRolesByStudy(ArachneUser user, Study study) {
 
-        return study != null ? getParticipantRoles(user.getId(), study) : new LinkedList<>();
+        List<ParticipantRole> participantRoles = new LinkedList<>();
+
+        if (study != null && study.getTenant().getId().equals(user.getActiveTenantId())) {
```
Comment (wivern) [inline src/main/java/com/odysseusinc/arachne/portal/service/impl/BaseArachneSecureServiceImpl.java]: Objects.equals looks slightly better?

File: src/main/java/com/odysseusinc/arachne/portal/service/impl/BaseUserServiceImpl.java [hunk 7]
```diff
@@ -800,16 +808,24 @@ public void indexAllBySolr()
...
         solrService.deleteByQuery(SolrServiceImpl.USER_COLLECTION, "*:*");
-        List<U> userList = getAllEnabled();
+        List<U> userList = getAllEnabledFromAllTenants();
         for (U user : userList) {
             indexBySolr(user);
         }
     }
 
-    public SearchResult<U> search(SolrQuery solrQuery) throws IOException, SolrServerException {
+    private SolrQuery addTenantFilter(SolrQuery solrQuery) throws NoSuchFieldException {
```
Comment (mpozhidaeva) [inline src/main/java/com/odysseusinc/arachne/portal/service/impl/BaseUserServiceImpl.java]: almost like addTenantFilter() in BaseDataSourceServiceImpl.java

File: src/main/java/com/odysseusinc/arachne/portal/service/domain/DomainObjectLoaderFactory.java [hunk 8]
```diff
@@ -43,7 +44,13 @@ public DomainObjectLoaderFactory(WebApplicationContext appContext, EntityManager
 
     public DomainObjectLoader getDomainObjectLoader(Class domainClazz) {
 
-        DomainObjectLoader domainObjectLoader = new GenericDomainObjectLoader(domainClazz);
+        DomainObjectLoader domainObjectLoader;
+
+        if (User.class.isAssignableFrom(domainClazz))
```
Comment (mpozhidaeva) [inline src/main/java/com/odysseusinc/arachne/portal/service/domain/DomainObjectLoaderFactory.java]: brackets or ternary operator

File: src/main/resources/db/migration/V20180206144043__tenant_domains.sql [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+ALTER TABLE studies RENAME TO studies_data;
+
+-- Since the view is simple, it is automatically becomes updatable one
+CREATE OR REPLACE VIEW studies AS
+SELECT *
+FROM studies_data
+WHERE tenant_id = current_setting('app.tenant_id')::BIGINT;
+
+
+-- 1) Dependant views reference the "studies" table by OID, not table name,
```
Comment (mpozhidaeva) [inline src/main/resources/db/migration/V20180206144043__tenant_domains.sql]: "dependent"?

**Pull request merged**